File: npm/server/src/validation.ts
/**
 * Input validation based on schema directives.
 *
 * Supports: @email, @minLength, @maxLength, @min, @max, @pattern, @url, @trim, etc.
 */

import { validationError, type ValidationError } from './types';

/**
 * Validation result.
 */
export type ValidationResult =
  | { valid: true; value: unknown }
  | { valid: false; errors: ValidationError[] };

/**
 * Validator function type.
 */
export type ValidatorFn = (
  value: unknown,
  args: Record<string, unknown>,
  fieldName: string
) => ValidationError | null;

/**
 * Built-in validators.
 */
export const validators: Record<string, ValidatorFn> = {
  /**
   * Validates email format.
   */
  email: (value, _args, fieldName) => {
    if (typeof value !== 'string') {
      return validationError(fieldName, '@email', 'Value must be a string');
    }
    // Simple email regex - for production use a more robust solution
    const emailRegex = /^[^\s@]+@[^\s@]+\.[^\s@]+$/;
    if (!emailRegex.test(value)) {
      return validationError(fieldName, '@email', 'Invalid email format');
    }
    return null;
  },

  /**
   * Validates minimum string length.
   */
  minLength: (value, args, fieldName) => {
    if (typeof value !== 'string') {
      return validationError(fieldName, '@minLength', 'Value must be a string');
    }
    const min = args.value as number;
    if (value.length < min) {
      return validationError(
        fieldName,
        '@minLength',
        `String must be at least ${min} characters`
      );
    }
    return null;
  },

  /**
   * Validates maximum string length.
   */
  maxLength: (value, args, fieldName) => {
    if (typeof value !== 'string') {
      return validationError(fieldName, '@maxLength', 'Value must be a string');
    }
    const max = args.value as number;
    if (value.length > max) {
      return validationError(
        fieldName,
        '@maxLength',
        `String must be at most ${max} characters`
      );
    }
    return null;
  },

  /**
   * Validates minimum numeric value.
   */
  min: (value, args, fieldName) => {
    if (typeof value !== 'number') {
      return validationError(fieldName, '@min', 'Value must be a number');
    }
    const min = args.value as number;
    if (value < min) {
      return validationError(
        fieldName,
        '@min',
        `Value must be at least ${min}`
      );
    }
    return null;
  },

  /**
   * Validates maximum numeric value.
   */
  max: (value, args, fieldName) => {
    if (typeof value !== 'number') {
      return validationError(fieldName, '@max', 'Value must be a number');
    }
    const max = args.value as number;
    if (value > max) {
      return validationError(
        fieldName,
        '@max',
        `Value must be at most ${max}`
      );
    }
    return null;
  },

  /**
   * Validates against a regex pattern.
   */
  pattern: (value, args, fieldName) => {
    if (typeof value !== 'string') {
      return validationError(fieldName, '@pattern', 'Value must be a string');
    }
    const pattern = new RegExp(args.regex as string);
    if (!pattern.test(value)) {
      const message = (args.message as string) ?? 'Value does not match required pattern';
      return validationError(fieldName, '@pattern', message);
    }
    return null;
  },

  /**
   * Validates URL format.
   */
  url: (value, _args, fieldName) => {
    if (typeof value !== 'string') {
      return validationError(fieldName, '@url', 'Value must be a string');
    }
    try {
      new URL(value);
      return null;
    } catch {
      return validationError(fieldName, '@url', 'Invalid URL format');
    }
  },

  /**
   * Validates maximum array items.
   */
  maxItems: (value, args, fieldName) => {
    if (!Array.isArray(value)) {
      return validationError(fieldName, '@maxItems', 'Value must be an array');
    }
    const max = args.value as number;
    if (value.length > max) {
      return validationError(
        fieldName,
        '@maxItems',
        `Array must have at most ${max} items`
      );
    }
    return null;
  },

  /**
   * Validates minimum array items.
   */
  minItems: (value, args, fieldName) => {
    if (!Array.isArray(value)) {
      return validationError(fieldName, '@minItems', 'Value must be an array');
    }
    const min = args.value as number;
    if (value.length < min) {
      return validationError(
        fieldName,
        '@minItems',
        `Array must have at least ${min} items`
      );
    }
    return null;
  },

  /**
   * Validates that value is positive.
   */
  positive: (value, _args, fieldName) => {
    if (typeof value !== 'number') {
      return validationError(fieldName, '@positive', 'Value must be a number');
    }
    if (value <= 0) {
      return validationError(fieldName, '@positive', 'Value must be positive');
    }
    return null;
  },

  /**
   * Validates that value is non-negative.
   */
  nonNegative: (value, _args, fieldName) => {
    if (typeof value !== 'number') {
      return validationError(fieldName, '@nonNegative', 'Value must be a number');
    }
    if (value < 0) {
      return validationError(fieldName, '@nonNegative', 'Value must be non-negative');
    }
    return null;
  },
};

/**
 * Transformers that modify input values.
 */
export const transformers: Record<string, (value: unknown) => unknown> = {
  /**
   * Trims whitespace from strings.
   */
  trim: (value) => {
    if (typeof value === 'string') {
      return value.trim();
    }
    return value;
  },

  /**
   * Converts string to lowercase.
   */
  lowercase: (value) => {
    if (typeof value === 'string') {
      return value.toLowerCase();
    }
    return value;
  },

  /**
   * Converts string to uppercase.
   */
  uppercase: (value) => {
    if (typeof value === 'string') {
      return value.toUpperCase();
    }
    return value;
  },
};

/**
 * Validation rule.
 */
export interface ValidationRule {
  readonly directive: string;
  readonly args: Record<string, unknown>;
}

/**
 * Field validation configuration.
 */
export interface FieldValidation {
  readonly fieldName: string;
  readonly rules: ValidationRule[];
  readonly transforms: string[];
}

/**
 * Validates a single field.
 */
export function validateField(
  value: unknown,
  config: FieldValidation
): ValidationResult {
  const errors: ValidationError[] = [];
  let transformedValue = value;

  // Apply transforms first
  for (const transform of config.transforms) {
    const transformer = transformers[transform];
    if (transformer) {
      transformedValue = transformer(transformedValue);
    }
  }

  // Skip validation for null/undefined (use @required for that)
  if (transformedValue === null || transformedValue === undefined) {
    return { valid: true, value: transformedValue };
  }

  // Run validators
  for (const rule of config.rules) {
    const validator = validators[rule.directive];
    if (validator) {
      const error = validator(transformedValue, rule.args, config.fieldName);
      if (error) {
        errors.push(error);
      }
    }
  }

  if (errors.length > 0) {
    return { valid: false, errors };
  }

  return { valid: true, value: transformedValue };
}

/**
 * Validates an entire input object.
 */
export function validateInput(
  input: Record<string, unknown>,
  fields: FieldValidation[]
): ValidationResult {
  const errors: ValidationError[] = [];
  const validatedInput: Record<string, unknown> = { ...input };

  for (const field of fields) {
    const value = input[field.fieldName];
    const result = validateField(value, field);

    if (!result.valid) {
      errors.push(...result.errors);
    } else {
      validatedInput[field.fieldName] = result.value;
    }
  }

  if (errors.length > 0) {
    return { valid: false, errors };
  }

  return { valid: true, value: validatedInput };
}

/**
 * Creates a validation middleware for resolvers.
 */
export function createValidationMiddleware(fields: FieldValidation[]) {
  return function validateArgs<TArgs extends Record<string, unknown>>(
    args: TArgs
  ): { valid: true; args: TArgs } | { valid: false; error: ValidationError } {
    const result = validateInput(args, fields);

    if (!result.valid) {
      return { valid: false, error: result.errors[0] };
    }

    return { valid: true, args: result.value as TArgs };
  };
}

/**
 * Registers a custom validator.
 */
export function registerValidator(name: string, fn: ValidatorFn): void {
  validators[name] = fn;
}

/**
 * Registers a custom transformer.
 */
export function registerTransformer(
  name: string,
  fn: (value: unknown) => unknown
): void {
  transformers[name] = fn;
}

File: sdk/rust/src/validation.rs
//! Input validation support for Better GraphQL.
//!
//! This module provides validation primitives that correspond to BGQL's
//! validation directives:
//!
//! - `@minLength`, `@maxLength` - String length constraints
//! - `@min`, `@max` - Numeric range constraints
//! - `@email`, `@url` - Format validators
//! - `@pattern` - Regex pattern matching
//! - `@trim`, `@lowercase`, `@sanitize` - Input transformations
//!
//! # Example
//!
//! ```ignore
//! use bgql_sdk::validation::{Validate, ValidationResult, Validator};
//!
//! // Simple validation
//! let mut validator = Validator::new();
//! validator.validate_string("name", name, |v| v.min_length(2).max_length(100));
//! validator.validate_string("email", email, |v| v.email());
//! let result = validator.finish();
//! ```

use serde::{Deserialize, Serialize};
use std::fmt;

/// A validation error for a specific field.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ValidationError {
    /// The field that failed validation.
    pub field: String,
    /// Human-readable error message.
    pub message: String,
    /// Machine-readable error code.
    pub code: ValidationErrorCode,
    /// The constraint that was violated (for min/max etc).
    #[serde(skip_serializing_if = "Option::is_none")]
    pub constraint: Option<String>,
}

impl ValidationError {
    /// Creates a new validation error.
    pub fn new(
        field: impl Into<String>,
        code: ValidationErrorCode,
        message: impl Into<String>,
    ) -> Self {
        Self {
            field: field.into(),
            message: message.into(),
            code,
            constraint: None,
        }
    }

    /// Sets the constraint value.
    pub fn with_constraint(mut self, constraint: impl Into<String>) -> Self {
        self.constraint = Some(constraint.into());
        self
    }
}

impl fmt::Display for ValidationError {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{}: {}", self.field, self.message)
    }
}

impl std::error::Error for ValidationError {}

/// Validation error codes.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "SCREAMING_SNAKE_CASE")]
pub enum ValidationErrorCode {
    /// Value is required but missing.
    Required,
    /// String is too short.
    MinLength,
    /// String is too long.
    MaxLength,
    /// Number is too small.
    Min,
    /// Number is too large.
    Max,
    /// Invalid email format.
    InvalidEmail,
    /// Invalid URL format.
    InvalidUrl,
    /// Pattern mismatch.
    PatternMismatch,
    /// Invalid format.
    InvalidFormat,
    /// Custom validation failed.
    Custom,
}

impl fmt::Display for ValidationErrorCode {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            Self::Required => write!(f, "REQUIRED"),
            Self::MinLength => write!(f, "MIN_LENGTH"),
            Self::MaxLength => write!(f, "MAX_LENGTH"),
            Self::Min => write!(f, "MIN"),
            Self::Max => write!(f, "MAX"),
            Self::InvalidEmail => write!(f, "INVALID_EMAIL"),
            Self::InvalidUrl => write!(f, "INVALID_URL"),
            Self::PatternMismatch => write!(f, "PATTERN_MISMATCH"),
            Self::InvalidFormat => write!(f, "INVALID_FORMAT"),
            Self::Custom => write!(f, "CUSTOM"),
        }
    }
}

/// Result type for validation operations.
pub type ValidationResult<T> = Result<T, ValidationErrors>;

/// Collection of validation errors.
#[derive(Debug, Clone, Default, Serialize, Deserialize)]
pub struct ValidationErrors {
    /// List of validation errors.
    pub errors: Vec<ValidationError>,
}

impl ValidationErrors {
    /// Creates an empty error collection.
    pub fn new() -> Self {
        Self { errors: Vec::new() }
    }

    /// Creates errors from a single error.
    pub fn single(error: ValidationError) -> Self {
        Self {
            errors: vec![error],
        }
    }

    /// Adds an error.
    pub fn push(&mut self, error: ValidationError) {
        self.errors.push(error);
    }

    /// Returns true if there are any errors.
    pub fn has_errors(&self) -> bool {
        !self.errors.is_empty()
    }

    /// Returns true if there are no errors.
    pub fn is_valid(&self) -> bool {
        self.errors.is_empty()
    }

    /// Returns errors for a specific field.
    pub fn for_field(&self, field: &str) -> Vec<&ValidationError> {
        self.errors.iter().filter(|e| e.field == field).collect()
    }

    /// Merges another set of errors.
    pub fn merge(&mut self, other: ValidationErrors) {
        self.errors.extend(other.errors);
    }

    /// Converts to a GraphQL-compatible error response.
    pub fn to_graphql_error(&self) -> serde_json::Value {
        serde_json::json!({
            "__typename": "ValidationError",
            "message": self.to_string(),
            "code": "VALIDATION_ERROR",
            "field": self.errors.first().map(|e| &e.field),
            "constraint": self.errors.first().and_then(|e| e.constraint.as_ref()),
        })
    }
}

impl fmt::Display for ValidationErrors {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        let messages: Vec<_> = self.errors.iter().map(|e| e.to_string()).collect();
        write!(f, "{}", messages.join("; "))
    }
}

impl std::error::Error for ValidationErrors {}

impl From<ValidationError> for ValidationErrors {
    fn from(error: ValidationError) -> Self {
        Self::single(error)
    }
}

/// Trait for validatable types.
pub trait Validate {
    /// Validates the value and returns errors if invalid.
    fn validate(&self) -> ValidationResult<()>;

    /// Returns true if the value is valid.
    fn is_valid(&self) -> bool {
        self.validate().is_ok()
    }
}

/// Builder for validating input fields.
#[derive(Debug, Default)]
pub struct Validator {
    errors: ValidationErrors,
}

impl Validator {
    /// Creates a new validator.
    pub fn new() -> Self {
        Self::default()
    }

    /// Validates a string field with the given checks.
    pub fn string(&mut self, name: &str, value: &str) -> StringValidator<'_> {
        StringValidator {
            errors: &mut self.errors,
            name: name.to_string(),
            value: value.to_string(),
        }
    }

    /// Validates a numeric field with the given checks.
    pub fn number<T: NumericValue>(&mut self, name: &str, value: T) -> NumericValidator<'_, T> {
        NumericValidator {
            errors: &mut self.errors,
            name: name.to_string(),
            value,
        }
    }

    /// Adds a custom validation error.
    pub fn add_error(&mut self, field: &str, code: ValidationErrorCode, message: &str) {
        self.errors.push(ValidationError::new(field, code, message));
    }

    /// Finishes validation and returns the result.
    pub fn finish(self) -> ValidationResult<()> {
        if self.errors.has_errors() {
            Err(self.errors)
        } else {
            Ok(())
        }
    }
}

/// Validator for string fields.
pub struct StringValidator<'a> {
    errors: &'a mut ValidationErrors,
    name: String,
    value: String,
}

impl<'a> StringValidator<'a> {
    /// Validates minimum length.
    pub fn min_length(self, min: usize) -> Self {
        if self.value.len() < min {
            self.errors.push(
                ValidationError::new(
                    &self.name,
                    ValidationErrorCode::MinLength,
                    format!("must be at least {} characters", min),
                )
                .with_constraint(min.to_string()),
            );
        }
        self
    }

    /// Validates maximum length.
    pub fn max_length(self, max: usize) -> Self {
        if self.value.len() > max {
            self.errors.push(
                ValidationError::new(
                    &self.name,
                    ValidationErrorCode::MaxLength,
                    format!("must be at most {} characters", max),
                )
                .with_constraint(max.to_string()),
            );
        }
        self
    }

    /// Validates email format.
    pub fn email(self) -> Self {
        if !is_valid_email(&self.value) {
            self.errors.push(ValidationError::new(
                &self.name,
                ValidationErrorCode::InvalidEmail,
                "must be a valid email address",
            ));
        }
        self
    }

    /// Validates URL format.
    pub fn url(self) -> Self {
        if !is_valid_url(&self.value) {
            self.errors.push(ValidationError::new(
                &self.name,
                ValidationErrorCode::InvalidUrl,
                "must be a valid URL",
            ));
        }
        self
    }

    /// Validates against a regex pattern.
    pub fn pattern(self, pattern: &str) -> Self {
        if let Ok(re) = regex::Regex::new(pattern) {
            if !re.is_match(&self.value) {
                self.errors.push(
                    ValidationError::new(
                        &self.name,
                        ValidationErrorCode::PatternMismatch,
                        "does not match the required pattern",
                    )
                    .with_constraint(pattern.to_string()),
                );
            }
        }
        self
    }

    /// Validates that the value is not empty after trimming.
    pub fn required(self) -> Self {
        if self.value.trim().is_empty() {
            self.errors.push(ValidationError::new(
                &self.name,
                ValidationErrorCode::Required,
                "is required",
            ));
        }
        self
    }
}

/// Trait for numeric values.
pub trait NumericValue: PartialOrd + Copy + fmt::Display {}
impl NumericValue for i32 {}
impl NumericValue for i64 {}
impl NumericValue for u32 {}
impl NumericValue for u64 {}
impl NumericValue for f32 {}
impl NumericValue for f64 {}

/// Validator for numeric fields.
pub struct NumericValidator<'a, T: NumericValue> {
    errors: &'a mut ValidationErrors,
    name: String,
    value: T,
}

impl<'a, T: NumericValue> NumericValidator<'a, T> {
    /// Validates minimum value.
    pub fn min(self, min: T) -> Self {
        if self.value < min {
            self.errors.push(
                ValidationError::new(
                    &self.name,
                    ValidationErrorCode::Min,
                    format!("must be at least {}", min),
                )
                .with_constraint(min.to_string()),
            );
        }
        self
    }

    /// Validates maximum value.
    pub fn max(self, max: T) -> Self {
        if self.value > max {
            self.errors.push(
                ValidationError::new(
                    &self.name,
                    ValidationErrorCode::Max,
                    format!("must be at most {}", max),
                )
                .with_constraint(max.to_string()),
            );
        }
        self
    }
}

/// Input transformation helpers.
pub mod transform {
    /// Trims whitespace from a string.
    pub fn trim(s: &str) -> String {
        s.trim().to_string()
    }

    /// Converts string to lowercase.
    pub fn lowercase(s: &str) -> String {
        s.to_lowercase()
    }

    /// Converts string to uppercase.
    pub fn uppercase(s: &str) -> String {
        s.to_uppercase()
    }

    /// Sanitizes HTML content.
    pub fn sanitize_html(s: &str) -> String {
        s.replace('<', "&lt;")
            .replace('>', "&gt;")
            .replace('"', "&quot;")
            .replace('\'', "&#x27;")
    }

    /// Applies trim and lowercase.
    pub fn normalize_email(s: &str) -> String {
        lowercase(&trim(s))
    }
}

/// Simple email validation (basic check).
fn is_valid_email(email: &str) -> bool {
    let trimmed = email.trim();
    if trimmed.is_empty() {
        return false;
    }

    // Basic email validation
    let parts: Vec<&str> = trimmed.split('@').collect();
    if parts.len() != 2 {
        return false;
    }

    let local = parts[0];
    let domain = parts[1];

    if local.is_empty() || domain.is_empty() {
        return false;
    }

    if !domain.contains('.') {
        return false;
    }

    let domain_parts: Vec<&str> = domain.split('.').collect();
    if domain_parts.iter().any(|p| p.is_empty()) {
        return false;
    }

    true
}

/// Simple URL validation.
fn is_valid_url(url: &str) -> bool {
    let trimmed = url.trim();
    if trimmed.is_empty() {
        return false;
    }

    // Basic URL validation
    (trimmed.starts_with("http://") || trimmed.starts_with("https://")) && trimmed.len() > 10
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_string_validation() {
        let mut validator = Validator::new();
        validator.string("name", "Al").min_length(3).max_length(100);
        let result = validator.finish();

        assert!(result.is_err());
        let errors = result.unwrap_err();
        assert_eq!(errors.errors.len(), 1);
        assert_eq!(errors.errors[0].code, ValidationErrorCode::MinLength);
    }

    #[test]
    fn test_email_validation() {
        // Valid email
        let mut validator = Validator::new();
        validator.string("email", "test@example.com").email();
        let result = validator.finish();
        assert!(result.is_ok());

        // Invalid email
        let mut validator = Validator::new();
        validator.string("email", "not-an-email").email();
        let result = validator.finish();
        assert!(result.is_err());
    }

    #[test]
    fn test_numeric_validation() {
        let mut validator = Validator::new();
        validator.number("age", 15).min(18).max(120);
        let result = validator.finish();

        assert!(result.is_err());
        let errors = result.unwrap_err();
        assert_eq!(errors.errors[0].code, ValidationErrorCode::Min);
    }

    #[test]
    fn test_multiple_fields() {
        let mut validator = Validator::new();
        validator.string("name", "").required().min_length(2);
        validator.string("email", "invalid").email();
        let result = validator.finish();

        assert!(result.is_err());
        let errors = result.unwrap_err();
        assert!(errors.errors.len() >= 2);
    }

    #[test]
    fn test_transform() {
        assert_eq!(transform::trim("  hello  "), "hello");
        assert_eq!(transform::lowercase("HELLO"), "hello");
        assert_eq!(transform::normalize_email("  TEST@Example.COM  "), "test@example.com");
        assert_eq!(transform::sanitize_html("<script>alert('xss')</script>"),
            "&lt;script&gt;alert(&#x27;xss&#x27;)&lt;/script&gt;");
    }

    #[test]
    fn test_validation_errors_to_graphql() {
        let errors = ValidationErrors::single(
            ValidationError::new("email", ValidationErrorCode::InvalidEmail, "Invalid email")
                .with_constraint("email format"),
        );

        let json = errors.to_graphql_error();
        assert_eq!(json["__typename"], "ValidationError");
        assert_eq!(json["field"], "email");
    }
}
